pyiron_base/generic/datacontainer.py

```diff
@@ -0,0 +1,730 @@
+                data = hdf_group["data"]
+                if "read_only" in hdf_group.list_nodes():
+                    read_only = hdf_group["read_only"]
+                else:
+                    read_only = False
+        else:
+            data = hdf["data"]
+            if "read_only" in hdf.list_nodes():
+                read_only = hdf["read_only"]
+            else:
+                read_only = False
+
+        self.clear()
+
+        self.update(data, wrap=True)
+        self.read_only = bool(read_only)
+
+    def nodes(self):
+        """
+        Iterator over keys to terminal nodes.
+
+        Returns:
+            :class:`list`: list of keys to normal values.
+        """
+        for k, v in self.items():
+            if not isinstance(v, DataContainer):
+                yield k
+
+    def list_nodes(self):
+        """
+        Return a list of keys to terminal nodes.
+
+        Returns:
+            :class:`list`: list of keys to normal values.
+        """
+        return list(self.nodes())
+
+    def groups(self):
+        """
+        Return a list of keys to nested containers.
+
+        Returns:
+            :class:`list`: list of all keys to elements of :class:`DataContainer`.
+        """
+        for k, v in self.items():
+            if isinstance(v, DataContainer):
+                yield k
+
+    def list_groups(self):
+        """
+        Return a list of keys to nested containers.
+
+        Returns:
+            :class:`list`: list of all keys to elements of :class:`DataContainer`.
+        """
+        return list(self.groups())
+
+    def read(self, file_name, wrap=False):
+        """
+        Parse file as dictionary and add its keys to this container.
+
+        Only yaml (*.yml, *.yaml) and xml (*.xml) files are supported.
+
+        Errors during reading of the files generate a warning, but leave the container unchanged.
+
+        Args:
+            file_name(str): path to the input file
+            wrap(bool), if set to true will wrap the inputed data itself as a datacontainer inside the datacontainer
+
+        Raises:
+            :class:`ValueError`: if file extension doesn't match one of the supported ones
+        """
+        ext = os.path.splitext(file_name)[1]
+        try:
+            parser = {
+                    'yml': _parse_yaml, 'yaml': _parse_yaml,
+                    'xml': _parse_xml
+            }[ext]
+        except KeyError:
+            raise ValueError("The input file is not supported; expected *.yml, *.yaml, or *.xml") from None
+
+        self.update(parser(file_name), wrap=wrap)
+
+    def write_yml(self, file_name):
+        with open(file_name, 'w') as output:
+            yaml.dump(self, output, default_flow_style=False)
+
+    def write_xml(self, file_name, attr_flag=False):
+        xml_data = dicttoxml(self, attr_type=attr_flag)
+        xmlfile = open(file_name, "w")
+        xmlfile.write(parseString(xml_data).toprettyxml())
+        xmlfile.close()
```

Comment:
Does this work with `with`?
```suggestion
        with open(file_name, "w") as xmlfile:
            xmlfile.write(parseString(xml_data).toprettyxml())
```